### PR TITLE
Provide Info about new connection to OnInit method

### PR DESCRIPTION
**Pull request by stanowawi**
Intention for new Feature:
The ODPI Driver provides information if the connection has been newly created, or it is taken from the connection pool ( see doc for dpiConnCreateParams.outNewSession) 
It is really useful to have this information by the OnInit method, to be able to differentiate between  connection from the pool or new connection.
Change:
- get and forward the "outNewSession" info as a return value by createConn & acquireConn  methods (newConnectionCreated true = new connection / false = pooled connection)
- Forward this info to the OnInit Method as a  new context tag. This solution with the context tag doesn't change OnInit signature = is  backwards compatible 
- Create new dsn.OnInitNewConnectionType / OnInitNewConnectionKey to provide unique context key for all users. 

 The whole change is backwards compatible.

### CODE REVIEW

**Comment by tgulacsi** [comment]
See defb84496261a07de2829aa84e7ec66264e24065 and #218, #219 - that outNewSession lied reproducibly.

**Comment by stanowawi** [comment]
I'm trying to reproduce the issue with the test out of #219 and my modification like this:
```
oraDSN.OnInit = func(ctx context.Context, conn driver.ConnPrepareContext) error {

newConFromOCI := ctx.Value(dsn.OnInitNewConnectionKey).(bool)
if newConFromOCI {
	qry := "ALTER SESSION SET NLS_DATE_FORMAT='YYYY-MM-DD HH24.MI.SS' NLS_TIMESTAMP_FORMAT='YYYY-MM-DD HH24.MI.SS.FF' NLS_TIMESTAMP_TZ_FORMAT='YYYY-MM-DD HH24.MI.SS.FF TZR' NLS_TIME_FORMAT='HH24.MI.SS.FF' NLS_TIME_TZ_FORMAT='HH24.MI.SS.FF TZR'"
	st, err := conn.PrepareContext(ctx, qry)
	if err == nil {
		_, err = st.(driver.StmtExecContext).ExecContext(ctx, nil)
		st.Close()
	}
....
```
I run it lot of times with the instantclient_19_16, instantclient_12_2 and instantclient_21_8, but. I can see the logic works, but I'm getting no issue :(
Really strange, do you have some recommendation?

File: conn.go [hunk 1]
```diff
@@ -1107,13 +1107,13 @@ func (c *conn) ResetSession(ctx context.Context) error {
 		// Just release
 		_ = c.closeNotLocking()
 	}
-	dpiConn, err := c.drv.acquireConn(pool, P)
+	dpiConn, newConnectionCreated, err := c.drv.acquireConn(pool, P)
```
Comment (tgulacsi) [inline conn.go:1110]: User shorter names, like "isNew"

File: conn.go [hunk 2]
```diff
@@ -1107,13 +1107,13 @@ func (c *conn) ResetSession(ctx context.Context) error {
...
 	c.mu.Unlock()
 	if err != nil {
 		return fmt.Errorf("%v: %w", err, driver.ErrBadConn)
 	}
 	c.dpiConn = dpiConn
-
+	ctx = context.WithValue(ctx, dsn.OnInitNewConnectionKey, newConnectionCreated)
```
Comment (tgulacsi) [inline conn.go:1116]: Why do we put this information into the context?

File: drv.go [hunk 3]
```diff
@@ -369,15 +369,16 @@ func (d *drv) initCommonCreateParams(P *C.dpiCommonCreateParams, enableEvents bo
...
+// second return value: true = connection is new / false = connection is from pool
+func (d *drv) createConn(pool *connPool, P commonAndConnParams) (*conn, bool, error) {
 	// initialize driver, if necessary
 	if err := d.init(P.ConfigDir, P.LibDir); err != nil {
-		return nil, err
+		return nil, false, err
 	}
 
-	dc, err := d.acquireConn(pool, P)
+	dc, newConnectionCreated, err := d.acquireConn(pool, P)
```
Comment (tgulacsi) [inline drv.go:379]: isNew

File: drv.go [hunk 4]
```diff
@@ -396,10 +397,10 @@ func (d *drv) createConn(pool *connPool, P commonAndConnParams) (*conn, error) {
 			c.params.Username = pool.params.Username
 		}
 	}
-	ctx, cancel := context.WithTimeout(context.Background(), nvlD(c.params.WaitTimeout, time.Minute))
+	ctx, cancel := context.WithTimeout(context.WithValue(context.Background(), dsn.OnInitNewConnectionKey, newConnectionCreated), nvlD(c.params.WaitTimeout, time.Minute))
```
Comment (tgulacsi) [inline drv.go:400]: why?

File: drv.go [hunk 5]
```diff
@@ -571,13 +572,15 @@ func (d *drv) acquireConn(pool *connPool, P commonAndConnParams) (*C.dpiConn, er
...
+			return nil, false, fmt.Errorf("pool=%p stats=%s params=%+v: %w",
 				pool.dpiPool, stats, connCreateParams, err)
 		}
-		return nil, fmt.Errorf("user=%q standalone params=%+v: %w",
+		return nil, false, fmt.Errorf("user=%q standalone params=%+v: %w",
 			username, connCreateParams, err)
 	}
-	return dc, nil
+	//use the information from ODPI driver if new connection has been created or it is only pooled
+	newConnctionCreated := connCreateParams.outNewSession == 1
```
Comment (tgulacsi) [inline drv.go:582]: isNew

File: drv.go [hunk 6]
```diff
@@ -597,15 +600,15 @@ func (d *drv) createConnFromParams(ctx context.Context, P dsn.ConnectionParams)
 			return nil, err
 		}
 	}
-	conn, err := d.createConn(pool, commonAndConnParams{CommonParams: P.CommonParams, ConnParams: P.ConnParams})
+	conn, newConnectionCreated, err := d.createConn(pool, commonAndConnParams{CommonParams: P.CommonParams, ConnParams: P.ConnParams})
```
Comment (tgulacsi) [inline drv.go:603]: isNew

File: drv.go [hunk 7]
```diff
@@ -597,15 +600,15 @@ func (d *drv) createConnFromParams(ctx context.Context, P dsn.ConnectionParams)
...
 	if err != nil {
 		return conn, err
 	}
 	onInit := getOnInit(&conn.params.CommonParams)
 	if onInit == nil {
 		return conn, err
 	}
-	ctx, cancel := context.WithTimeout(ctx, nvlD(conn.params.WaitTimeout, time.Minute))
+	ctx, cancel := context.WithTimeout(context.WithValue(ctx, dsn.OnInitNewConnectionKey, newConnectionCreated), nvlD(conn.params.WaitTimeout, time.Minute))
```
Comment (tgulacsi) [inline drv.go:611]: ?

File: dsn/dsn.go [hunk 8]
```diff
@@ -45,6 +45,13 @@ const (
 	DefaultStandaloneConnection = false
 )
 
+type OnInitNewConnectionType string
```
Comment (tgulacsi) [inline dsn/dsn.go:48]: Why do we need this at all?
If we do, use 
type isNewCtxKey struct{}

**Review by tgulacsi — CHANGES_REQUESTED** [review]
Ok, let's try it, but make it OPT-IN with an Option!

**Comment by stanowawi** [comment]
Hello Tamás.
Do you mean to change OnInit method and add optional parameter? This is I wanted to avoid, because all projects, which use it, will brake like this:  https://go.dev/play/p/j62gJWH1Xl5
I can see following options:
1. change OnInit signature (add new optional param) = not compatible
2. provide the "isNew" information over  Context, as I did = not nice but compatibility is not broken
3. introduce new OnInitNew with isNew Information method and make a OnInit deprecated

Do you want to take option 1. ?

File: conn.go (hunk 2 above)
Comment (stanowawi) [reply conn.go:1116]: See comment bellow. I don't want to change OnInit method and cause compile errors in other projects, the I use Context to provide info.

File: drv.go (hunk 3 above)
Comment (stanowawi) [reply drv.go:379]: yes, you are right I'll change it

File: conn.go (hunk 2 above)
Comment (tgulacsi) [reply conn.go]: Why would OnInit need this info?
It should be called iff it is a new connection.

File: dsn/dsn.go (hunk 8 above)
Comment (tgulacsi) [reply dsn/dsn.go]: Yup, 
```type isNewCtxKey struct{}```
is a new dedicated type for it. Do NOT use a string !

File: conn.go (hunk 2 above)
Comment (stanowawi) [reply conn.go]: I reworked according your requirements. Parameter = OnInitNewCon if true , then OnInit is called only by the new connection.
Please check.

File: dsn/dsn.go [hunk 9]
```diff
@@ -56,6 +56,8 @@ type CommonParams struct {
 	OnInit func(context.Context, driver.ConnPrepareContext) error
 	// OnInitStmts are executed on session init, iff OnInit is nil.
 	OnInitStmts []string
+	// true: OnInit will be called only by the new session / false: OnInit will called by new or pooled connection
+	OnInitNewCon bool
```
Comment (tgulacsi) [inline dsn/dsn.go:60]: Where does this get a value?

File: drv.go [hunk 10]
```diff
@@ -597,20 +600,22 @@ func (d *drv) createConnFromParams(ctx context.Context, P dsn.ConnectionParams)
...
 	if onInit == nil {
 		return conn, err
 	}
-	ctx, cancel := context.WithTimeout(ctx, nvlD(conn.params.WaitTimeout, time.Minute))
-	err = onInit(ctx, conn)
-	cancel()
-	if err != nil {
-		conn.Close()
-		return nil, err
+	if !P.CommonParams.OnInitNewCon || isNew {
```
Comment (tgulacsi) [inline drv.go:611]: Turn this around:
```
if P.CommonParams.OnInitNewCon && !isNew {
    return conn, nil
}
```

**Review by tgulacsi — CHANGES_REQUESTED** [review]

**Review by tgulacsi — CHANGES_REQUESTED** [review]
Sorry, one last stylistic change, please:
CommonParams.InitOnNewConn
and call the new param
"initOnNewConnection"
Thanks!

**Comment by stanowawi** [comment]
no problem 
the rename is done, please check. Thank you too.

**Pull request merged**